rust/crd/src/resource.rs

```diff
@@ -1,28 +1,75 @@
 use crate::storage;
+use crate::{DruidCluster, DruidRole};
 use lazy_static::lazy_static;
-use snafu::Snafu;
+use snafu::{ResultExt, Snafu};
+use stackable_operator::config::fragment;
+use stackable_operator::role_utils::RoleGroupRef;
 use stackable_operator::{
-    commons::resources::{CpuLimits, MemoryLimits, NoRuntimeLimits, Resources},
+    commons::resources::{
+        CpuLimitsFragment, MemoryLimits, MemoryLimitsFragment, NoRuntimeLimits,
+        NoRuntimeLimitsFragment, Resources, ResourcesFragment,
+    },
     config::merge::Merge,
     k8s_openapi::{
         api::core::v1::ResourceRequirements, apimachinery::pkg::api::resource::Quantity,
     },
 };
 use strum::{EnumDiscriminants, IntoStaticStr};
 
+/// This Error cannot derive PartialEq because fragment::ValidationError doesn't derive it
+#[derive(Snafu, Debug, EnumDiscriminants)]
+#[strum_discriminants(derive(IntoStaticStr))]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
+    #[snafu(display("no resources available for merging"))]
+    NoResourcesToMerge,
+    #[snafu(display("cannot merge storage types of different roles"))]
+    IncompatibleStorageMerging,
+    #[snafu(display("failed to validate resources"))]
+    ResourceValidation { source: fragment::ValidationError },
+    #[snafu(display("failed to merge resources for {rolegroup_ref}"))]
+    ResourcesMerge {
+        #[snafu(source(from(Error, Box::new)))]
+        source: Box<Error>,
+        rolegroup_ref: RoleGroupRef<DruidCluster>,
+    },
+}
+
+/// The sole puprose of this enum is to handle merging. It's needed because currently
+/// the operator-rs 0.26.1 doesn't handle fragment enum merging.
+#[derive(Debug, Clone, PartialEq)]
+enum RoleResourceFragment {
+    DruidFragment(ResourcesFragment<storage::DruidStorage, NoRuntimeLimits>),
+    HistoricalFragment(ResourcesFragment<storage::HistoricalStorage, NoRuntimeLimits>),
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub enum RoleResource {
     Druid(Resources<storage::DruidStorage, NoRuntimeLimits>),
     Historical(Resources<storage::HistoricalStorage, NoRuntimeLimits>),
 }
 
+impl From<RoleResourceFragment> for RoleResource {
+    fn from(rrf: RoleResourceFragment) -> Self {
+        match rrf {
+            RoleResourceFragment::DruidFragment(fragment) => {
+                RoleResource::Druid(fragment::validate(fragment).unwrap()) // this succeeds because the actual validation has happened in try_merge
+            }
+            RoleResourceFragment::HistoricalFragment(fragment) => {
+                RoleResource::Historical(fragment::validate(fragment).unwrap())
+            }
+        }
+    }
```

Comment:
Do we still want to throw an FragmentValidationFailed error instead of the unwrap? I know it cannot fail now, but we dont know what changes in the future?

Reply:
Done